### ISSUE TITLE
Hubspot migration

### ISSUE THREAD

**Issue by cjyabraham** [LF-Engineering/lfevents#616]
From @ccraigross :
Install [HubSpot plugin](https://wordpress.org/plugins/leadin/)

Add this code to <head> if it is not already inserted by the plugin:
```
<!-- Start of HubSpot Embed Code -->
<script type="text/javascript" id="hs-script-loader" async defer src="//js.hs-scripts.com/8112310.js"></script>
<!-- End of HubSpot Embed Code -->  
```

**Comment by thetwopct** [comment]
Using the HubSpot for WordPress Shortcode, the following scripts are inserted in to the page:

Header
https://js.hscollectedforms.net/collectedforms.js
https://js.hs-banner.com/8112310.js
https://js.hs-analytics.net/analytics/1610564400000/8112310.js

Footer
https://js.hsforms.net/forms/v2.js
https://js.hs-scripts.com/8112310.js?integration=WordPress

This causes some slow down:

![2021-01-13-233825@2x](https://user-images.githubusercontent.com/10615884/104525829-f5dbb880-55f8-11eb-9a11-44aeefce1144.jpg)

By adding the following to Flying Scripts we can delay the scripts slightly on page load / until user interaction:

collectedforms.js
js.hs-banner.com
js.hs-analytics.net
js.hs-scripts.com
v2.js

Additionally I've added preconnects to help speed up that initial connection:

Preconnect
https://js.hscollectedforms.net
https://js.hs-banner.com
https://js.hs-analytics.net
https://js.hsforms.net
https://js.hs-scripts.com

It's also possible to add native lazy load to the iFrame via a filter on the_content

Result:

![2021-01-13-231912@2x](https://user-images.githubusercontent.com/10615884/104521915-c7101300-55f5-11eb-8ee3-53de9e6ebbd2.jpg)

**Comment by cjyabraham** [comment]
Do we want to use [the iframe solution](https://dev-lfeventsci.pantheonsite.io/about/contact-2/) or is there a good reason to [embed HTML directly in the page](https://dev-lfeventsci.pantheonsite.io/about/form-in-raw-html-format/)?  Can we get the loading spinner implemented in either case?

**Comment by cjyabraham** [comment]
**Flying scripts**
I suggest we disable this for the following pages where there's a form above-the-fold which would obviously be delayed:
- [contact us](https://events.linuxfoundation.org/about/contact/) 
- [visa request](https://events.linuxfoundation.org/about/visa-request/)
- [travel fund request](https://events.linuxfoundation.org/about/travel-fund-request/)

**Configurable forms**
For forms like [visa request](https://events.linuxfoundation.org/about/visa-request/) where the editor configures the events that are available in the drop-down, that drop-down should now be configured in HubSpot rather than in WordPress.  You can allow everyone to edit the particular property in HubSpot to add or remove values.

**Comment by cjyabraham** [comment]
Other tasks needed to dismantle current forms solution once all forms have been migrated to HubSpot:
- [ ] delete visa-request.php and set [visa request page](https://events.linuxfoundation.org/about/visa-request/) to use default page template
- [ ] delete sfmc-forms.js 
- [ ] delete travel-fund-request.php and delete [travel fund request page](https://events.linuxfoundation.org/about/travel-fund-request/)
- [ ] remove blurb saying "Forms on this site are protected by reCAPTCHA ..." from footer
- [ ] stop enqueuing recaptcha script
- [ ] dismantle the 3 Event meta fields that control inclusion on visa request and travel fund forms

**Comment by thetwopct** [comment]
For editor-friendliness I figure using the iFrame is better than using HTML. 

The HTML option will allow us to style the forms completely using our CSS but that also introduces possible issues where an editor launching a form with complex layout may not see the layout they require. With so many issues with forms recently, I would prefer them to have the control they want.

But it seems iFrame or HTML option can be set per form, so we could take control of the styling if we needed too (like if editors start using Comic Sans for the form font etc 😂)

Flying Scripts - I don't see an issue with disabling it on those pages.

A) But then if it will be disabled is it still worth doing the loading spinner?

B) For configurable forms like visa request, instead of a dropdown of events for selection that needs to be updated, would it be better to have one visa request form per event? I can see a list of events in Hubspot getting confusing quite quickly...

C) The newsletters are currently configured via the sidebar for displaying newsletter form on every child event page and for overriding text, I guess we'd need to remove / update that section too.

**Comment by thetwopct** [comment]
I've added a spinner using CSS, which is the same method as the Custom iFrame Block uses.

![2021-01-18-130903](https://user-images.githubusercontent.com/10615884/104919652-80703f00-598e-11eb-8df4-403b5baf5f71.gif)

The downside to this method is that it requires the form background to always be white. So this will not work on forms on a coloured background i.e

![2021-01-18-131127@2x](https://user-images.githubusercontent.com/10615884/104919766-b0b7dd80-598e-11eb-9e6e-6f889efad90e.jpg)

An alternative implementation is to try and have the loading style class be removed via jQuery upon iFrame completing loading.

Instead of applying the loading to all forms, we could offer loading/background color as an optional newsletter preference in the event settings.

**Comment by cjyabraham** [comment]
You raise some good points.  My responses:
- Agreed that we should go with iframe by default, however, I could see us using the raw HTML option for things like the cncf and lfph footer forms, where we want to style precisely in the way we currently have done
- (A) I take your point and I think perhaps we should just try without the loading spinner for now since we'll be configuring things so that any load is hardly noticeable; I think seeing a spinner is a bit tacky
- (B) Yes, any event that we host on the events site will have their own visa request form.  The generic form, however, is needed for events that are not hosted on the events site (there are a few which insist on using their own promotional site).  So we need to be able to support them with a form where users select the event from a list.  I think this might be preferable to spinning up a new form for every external event we support, but that is an option as well.
- (C) Dang, ya!  That's a lot of custom functionality we'll need to replace or lose... hmm.
- Regarding jank, I think CLS may not be measured if you delay the appearance of the form enough with flying scripts.  I think without any delay, if you try [on this page](https://dev-lfeventsci.pantheonsite.io/about/contact-2/), for example, you'll see CLS of 0.402.

**Comment by cjyabraham** [comment]
As discussed:
- keep the loading spinner for now so we can see what it looks like in practice
- we will check with Megan whether her team is comfortable editing the Event dropdown for generic forms in HubSpot or whether we should find a way to implement as it was, using the Event sidebar in wp-admin

**Comment by thetwopct** [comment]
As a quick proof of concept, if we need to dynamically insert values in to a form field after Hubspot has loaded this can be done.

![2021-01-19-151605](https://user-images.githubusercontent.com/10615884/105054213-a28abf80-5a69-11eb-8fb6-7219cec27474.gif)

This is changing a normal input field (and after adding code and refreshing) in to a select box.

```
window.addEventListener('load', function () {
$('.hs-country input').replaceWith('<select><option value="1">Event 1</option></select>');
});

```

**Comment by cjyabraham** [comment]
The form that we need to do this on is huge and so it should ideally be in an iframe rather than the raw html version.  I tried to get this to work in an iframe with this code:
```
window.addEventListener('load', function () {
	$('iframe#hs-form-iframe-0').contents().find('.hs-event select').replaceWith('<select><option value="1">Event 1</option></select>');
});
```
Unfortunately it doesn't seem to work.  Any luck with getting this to work in the iframe version of a HubSpot form?

**Comment by cjyabraham** [comment]
I dug into this a bit more, started with a select control which I then edited the value of one of the options.  The form ended up not accepting the manipulated value.  My guess is hubspot will only accept one of the values for the dropdown for which it has configured on the server, which makes sense from a security POV.  So I think we're left with Megan and crew setting up the values within HubSpot, which is not a terrible solution.

**Comment by cjyabraham** [comment]
I removed the loading spinner for now because I don't like the requirement of having a white background on forms.  I also don't think we'll necessarily need the spinner.

**Comment by thetwopct** [comment]
A select control would have defined values on the hubspot side so we would need to take over a text input and replace it with a select option. The value of our inserted select is then passed as text so would have no validation on hubspot side. 

Anyway, if we're using the iFrame embed then manipulation of content inside the iFrame is not possible (security / SOP). Only for the HTML option.

**Comment by cjyabraham** [comment]
That's a cool idea.  I actually did try to manipulate the iframe content from js on the parent page and it is possible since the iframe is just created dynamically and has no `src` param. I got to the point where about half the time I could manipulate the controls as it was loading and the rest of the time some race condition prevented it.  Perhaps if we persisted we could get something which worked, however, I feel this approach is quite brittle and I think having Megan edit a param in HubSpot is not too much to ask.
